### PR TITLE
Add debugging docs

### PR DESCRIPTION
**Pull request by mercedesb**
## What was the end-user or developer problem that led to this PR?

Figuring out how to set up test cases and debug for a new maintainer is hard and there aren't existing docs to refer to. I had a call with @indirect and he walked me through how he debugs the Bundler code. This doc is a starting point for how to debug various test cases.

## What is your fix for the problem, implemented in this PR?

- Add DEBUGGING.md to document how to debug Bundler

## Make sure the following tasks are checked

- [ ] Describe the problem / feature
- [ ] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [ ] Write code to solve the problem
- [ ] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by simi** [comment]
:thinking: Usually I just put `binding.irb` in test and run that one test file or case individually. For integration testing I just run `ruby setup.rb` to install dev version locally or even run `ruby -Ilib  bin/gem` locally to test `gem` command.

File: bundler/doc/development/DEBUGGING.md
Comment (deivid-rodriguez) [inline bundler/doc/development/DEBUGGING.md:9]: I mainly use `puts` debugging too! Sometimes printing `caller_locations(0).join("\n")` helps to see what the stacktrace is at a given point.

File: bundler/doc/development/DEBUGGING.md
Comment (deivid-rodriguez) [inline bundler/doc/development/DEBUGGING.md:31]: I normally point directly to my working copy with something like `/path/to/rubygems/bundle/exe/bundle`. These days once you land on a copy of Bundler, you never get out of it because we always use `require_relative`, so messing with the `$LOAD_PATH` should not be necessary.

**Review by deivid-rodriguez — COMMENTED** [review]
I added a couple of comments. Debugging tips never hurt, thanks for documenting this!

File: bundler/doc/development/DEBUGGING.md
Comment (mercedesb) [reply bundler/doc/development/DEBUGGING.md:9]: Wonderful, I'll add that to the list of examples

File: bundler/doc/development/DEBUGGING.md
Comment (mercedesb) [reply bundler/doc/development/DEBUGGING.md:31]: Ah so should we change the alias instructions in the setup doc?

File: bundler/doc/development/DEBUGGING.md
Comment (mercedesb) [reply bundler/doc/development/DEBUGGING.md:47]: Because `binding.irb` is a REPL and not a debugger 🤷‍♀️ It's nice, but you can't really interactive debug. I can add that to these docs as a tool to use though

**Comment by mercedesb** [comment]
> For integration testing I just run `ruby setup.rb` to install dev version locally or even run `ruby -Ilib  bin/gem` locally to test gem command.

> I normally point directly to my working copy with something like `/path/to/rubygems/bundle/exe/bundle`

It sounds like there are multiple ways to test locally which don't match up with what we currently recommend to new maintainers in [SETUP.md](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/SETUP.md). Is the alias a totally bad idea? And should we update that doc too?

File: bundler/doc/development/DEBUGGING.md
Comment (simi) [reply bundler/doc/development/DEBUGGING.md:47]: That's it. Some info is at https://github.com/rubygems/rubygems/blob/030e285881ac99608fe8a23121b80ce1ce14ed6e/bundler/doc/development/SETUP.md#debugging-with-pry. Instead of this "complex" setup, you can do just `binding.irb` and you have basic debug console. For debug console, you can use `binding.break` (you need to `require 'debug'` somehow from [debug](https://github.com/ruby/debug) gem, which is distributed with Ruby if I remember well and there is no need to install additional dependency).

File: bundler/doc/development/DEBUGGING.md
Comment (deivid-rodriguez) [reply bundler/doc/development/DEBUGGING.md:31]: I don't mind the current recommendation to set an alias, and it should also work just fine, I don't think you need to change anything. Just wanted to mention the way I do it :)

File: bundler/doc/development/DEBUGGING.md
Comment (simi) [reply bundler/doc/development/DEBUGGING.md:47]: You can install debug gem the same way you can install `pry-byebug` for older Rubies. It supports Ruby 2.6+. 

Anyway the whole discussion is good sign that this setup is not simple. Experienced developers can use their favourite tools, there is almost nothing specific. And for beginners it would be maybe better to point to some generic Ruby debugging tutorial where basics like puts debugging and real debuggers are explained. :thinking: 

On the other way, if you think @indirect this kind of guide is valuable in here, let's start somewhere. I'm ok with anything that works. I was just about share my experience.

Comment (mercedesb) [reply bundler/doc/development/DEBUGGING.md]: Based on this convo, I reorganized the doc slightly and tried to add just enough info to get new maintainers started but would like to link out to other places for more learning. Do we care who/what we link to? Here are some options (not in the doc yet)

- REPL debugging more info: https://st0012.dev/whats-new-in-ruby-3-2-irb
- Prints debugging more info: https://medium.com/@calielrushton/puts-prints-and-more-printing-to-the-console-in-ruby-7830cdaace0a
- Interactive debugging more info: https://github.com/ruby/debug#how-to-use

File: bundler/doc/development/DEBUGGING.md
Comment (simi) [inline bundler/doc/development/DEBUGGING.md:31]: Maybe `https://github.com/ruby/irb#the-bindingirb-breakpoint`?

File: bundler/doc/development/DEBUGGING.md
Comment (simi) [reply bundler/doc/development/DEBUGGING.md:60]: I think everyone have different debugging habits :). I was using `ruby -Ilib -s ./exe/bundle -- --version` and even today it is usually enough to just call `./exe/bundle`. 

But indeed, `dbundle` is mentioned at various places. It is ok to keep it in here as well.

**Review by simi — APPROVED** [review]

**Pull request merged**
